core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,166 @@
+import static org.apache.iceberg.TableProperties.COMMIT_MAX_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_MIN_RETRY_WAIT_MS_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES;
+import static org.apache.iceberg.TableProperties.COMMIT_NUM_RETRIES_DEFAULT;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS;
+import static org.apache.iceberg.TableProperties.COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT;
+
+class UpdateSnapshotReferencesOperation implements PendingUpdate<Map<String, SnapshotRef>> {
+
+  private final TableOperations ops;
+  private TableMetadata base;
+  private final Map<String, SnapshotRef> refsToUpdate;
+  private final Set<String> refsToRemove;
+
+  UpdateSnapshotReferencesOperation(TableOperations ops) {
+    this.ops = ops;
+    this.base = ops.current();
+    this.refsToUpdate = Maps.newHashMap();
+    this.refsToRemove = Sets.newHashSet();
+  }
+
+  @Override
+  public Map<String, SnapshotRef> apply() {
+    this.base = ops.refresh();
+    Map<String, SnapshotRef> refs = Maps.newHashMap();
+    TableMetadata.Builder updatedBuilder = TableMetadata.buildFrom(base);
+    for (String refName : refsToRemove) {
+      SnapshotRef ref = base.ref(refName);
+      ValidationException.check(ref != null, "Cannot remove nonexistent snapshot ref %s", refName);
+      if (ref.isBranch()) {
+        updatedBuilder.removeBranch(refName);
+      }
+      else {
+        updatedBuilder.removeTag(refName);
+      }
+    }
+    for (Map.Entry<String, SnapshotRef> refEntry : refsToUpdate.entrySet()) {
+      String name = refEntry.getKey();
+      SnapshotRef snapshotRef = refEntry.getValue();
+      updatedBuilder.createSnapshotRef(name, snapshotRef);
+    }
+    return refs;
+  }
+
+  @Override
+  public void commit() {
+    Tasks.foreach(ops)
+        .retry(base.propertyAsInt(COMMIT_NUM_RETRIES, COMMIT_NUM_RETRIES_DEFAULT))
+        .exponentialBackoff(
+            base.propertyAsInt(COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(taskOps -> {
+          Map<String, SnapshotRef> refs = apply();
+          TableMetadata.Builder updatedMetadata = TableMetadata.buildFrom(base);
+          updatedMetadata.
+          taskOps.commit(base, updated);
+        });
+  }
+
+  public UpdateSnapshotReferencesOperation createBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef branch = SnapshotRef.branchBuilder(snapshotId).build();
+    refsToUpdate.put(name, branch);
```

Comment:
This should check that no ref was returned by `put`. Otherwise this just overwrites an existing branch.

Reply:
Good point. If we do transaction.updateSNapshotReferenceOperations().createBranch("branch1", 1).createBranch("branch1", 2) this should fail, because independent commits would fail due to the ref already existing. To have equivalence with the chaining, it should also fail.